src/main/scala/cromwell/engine/workflow/WorkflowMetadataBuilder.scala

```diff
@@ -0,0 +1,164 @@
+package cromwell.engine.workflow
+
+import cromwell.engine
+import cromwell.engine.ExecutionIndex._
+import cromwell.engine._
+import cromwell.engine.backend.{Backend, CallMetadata, WorkflowLogs}
+import cromwell.engine.db.slick._
+import cromwell.engine.db.{CallStatus, ExecutionDatabaseKey}
+import cromwell.engine.finalcall.FinalCall._
+import cromwell.engine.workflow.WorkflowManagerActor._
+import cromwell.webservice._
+import org.joda.time.DateTime
+import spray.json._
+import wdl4s._
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.io.Source
+
+object WorkflowMetadataBuilder {
+
+  // TODO: This assertion could also be added to the db layer: "In the future I'll fail if the workflow doesn't exist"
+  def assertWorkflowExistence(id: WorkflowId, workflowState: Option[WorkflowState]): Future[Unit] = {
+    // Confirm the workflow exists by querying its state.  If no state is found the workflow doesn't exist.
+    workflowState match {
+      case None => Future.failed(new WorkflowNotFoundException(s"Workflow '$id' not found"))
+      case _ => Future.successful(Unit)
+    }
+  }
+
+  private def assertCallFqnWellFormed(descriptor: WorkflowDescriptor, callFqn: FullyQualifiedName): String = {
+    descriptor.namespace.resolve(callFqn) match {
+      case Some(c: Call) => c.unqualifiedName
+      case _ => throw new UnsupportedOperationException(
+        s"Expected a fully qualified name to have at least two parts but got $callFqn")
+    }
+  }
+
+  private def hasLogs(entries: Iterable[ExecutionDatabaseKey])(key: ExecutionDatabaseKey) = {
+    !key.fqn.isScatter && !key.isCollector(entries)
+  }
+
+  def workflowStdoutStderr(workflowId: WorkflowId, backend: Backend,
+                           workflowDescriptor: WorkflowDescriptor,
+                           executionStatuses: Map[ExecutionDatabaseKey, CallStatus]): WorkflowLogs = {
+    val statusMap = executionStatuses filterKeys {
+      !_.fqn.isFinalCall
+    } mapValues {
+      _.executionStatus
+    }
+
+    val sortedMap = statusMap.toSeq.sortBy(_._1.index)
+    val callsToPaths = for {
+      (key, status) <- sortedMap if !key.fqn.isFinalCall && hasLogs(statusMap.keys)(key)
+      callName = assertCallFqnWellFormed(workflowDescriptor, key.fqn)
+      call = workflowDescriptor.namespace.workflow.findCallByName(callName).get
+      callKey = BackendCallKey(call, key.index, key.attempt)
+      // TODO There should be an easier way than going as far as backend.bindCall just to retrieve stdout/err path
```

Comment:
yeah the pluggable backend version of this was a good bit less horrible, hopefully something like that will be coming here soon